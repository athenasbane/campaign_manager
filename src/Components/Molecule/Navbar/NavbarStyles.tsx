import Box from "@mui/material/Box";
import styled from "@mui/material/styles/styled";
import { keyframes } from "@mui/material/styles";

const sparkAnimation = keyframes`
  0%, 100% { opacity: 0; transform: scale(0); }
  50% { opacity: 1; transform: scale(1); }
`;

export const StyledNavbar = styled(Box)(({ theme }) => ({
  position: "relative",
  overflow: "hidden",
  width: "100vw",
  paddingTop: theme.spacing(8),
  paddingBottom: theme.spacing(4),
  backgroundColor: "#111",
}));

interface SparkProps {
  top: string;
  left: string;
  delay: string;
}

export const Spark = styled("span", {
  shouldForwardProp: (prop) => !["top", "left", "delay"].includes(prop as string),
})<SparkProps>(({ top, left, delay }) => ({
  position: "absolute",
  width: "6px",
  height: "6px",
  borderRadius: "50%",
  backgroundColor: "#00d4ff",
  boxShadow: "0 0 8px 4px rgba(0,212,255,0.8)",
  opacity: 0,
  top,
  left,
  animation: `${sparkAnimation} 2s linear infinite`,
  animationDelay: delay,
}));
