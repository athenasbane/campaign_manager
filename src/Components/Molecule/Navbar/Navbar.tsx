import Stack from "@mui/material/Stack";
import Button from "@mui/material/Button";
import Typography from "@mui/material/Typography";
import Box from "@mui/material/Box";
import { Link, useLocation } from "react-router-dom";
import ArrowBackIosIcon from "@mui/icons-material/ArrowBackIos";
import { useNavigate } from "react-router-dom";
import { useEffect, useRef, useMemo } from "react";
import { useAppDispatch } from "../../../hooks/store.hooks";
import { EnumLayout, setLayoutValue } from "../../../Store/slices/layout";
import { Spark, StyledNavbar } from "./NavbarStyles";

export interface INavbarProps {
  onMenuButtonClick: () => void;
}

export default function Navbar({ onMenuButtonClick }: INavbarProps) {
  const navigate = useNavigate();
  const location = useLocation();
  const navRef = useRef<null | HTMLDivElement>(null);
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (navRef.current) {
      const { height, width } = navRef.current.getBoundingClientRect();
      dispatch(
        setLayoutValue({
          component: EnumLayout.NavBar,
          details: { height, width },
        })
      );
    }
  });
  const sparks = useMemo(
    () =>
      Array.from({ length: 15 }).map(() => ({
        top: `${Math.random() * 100}%`,
        left: `${Math.random() * 100}%`,
        delay: `${Math.random() * 5}s`,
      })),
    []
  );

  return (
    <StyledNavbar ref={navRef}>
      {sparks.map((s, i) => (
        <Spark key={i} top={s.top} left={s.left} delay={s.delay} />
      ))}
      <Stack direction="row" alignItems="center">
        <Box sx={{ width: "25%", textAlign: "center" }}>
          {location.pathname !== "/" ? (
            <Button color="primary" onClick={() => navigate(-1)}>
              <ArrowBackIosIcon color="inherit" />
            </Button>
          ) : null}
        </Box>
        <Box sx={{ width: "50%" }}>
          <Link
            style={{
              textDecoration: "none",
              color: "white",
            }}
            to="/"
          >
            <Typography variant="h3" textAlign="center">
              Teratin
            </Typography>
          </Link>
        </Box>
        <Box sx={{ width: "25%", textAlign: "center" }}>
          <Button
            data-testid="menu__button"
            variant="contained"
            color="primary"
            onClick={onMenuButtonClick}
          >
            <Typography sx={{ color: "#ffff" }}>Menu</Typography>
          </Button>
        </Box>
      </Stack>
    </StyledNavbar>
  );
}
