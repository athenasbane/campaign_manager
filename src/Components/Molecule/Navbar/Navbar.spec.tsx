import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import Navbar from "Components/Molecule/Navbar/Navbar";

describe("Molecule - Navbar", () => {
  let onMenuButtonClickStub: jest.Mock;
  beforeEach(() => {
    onMenuButtonClickStub = jest.fn();
  });

  it("should match the snapshot", () => {
    const { container } = render(
      <Navbar onMenuButtonClick={onMenuButtonClickStub} />,
      {
        wrapper: BrowserRouter,
      }
    );

    expect(container).toMatchSnapshot();
  });
  it("should call stub when menu button clicked", async () => {
    render(<Navbar onMenuButtonClick={onMenuButtonClickStub} />, {
      wrapper: BrowserRouter,
    });

    userEvent.click(screen.queryByTestId("menu__button") as HTMLElement);

    expect(onMenuButtonClickStub).toHaveBeenCalled();
  });
});
