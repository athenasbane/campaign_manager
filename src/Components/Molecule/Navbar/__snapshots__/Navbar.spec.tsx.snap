// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecule - Navbar should match the snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-1i9rmbe"
  >
    <div
      class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-kgdgla-MuiGrid-root"
      />
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 css-1osj8n2-MuiGrid-root"
      >
        <a
          href="/"
          style="text-decoration: none; color: white;"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-149xwtl-MuiTypography-root"
          >
            Tordenhelm
          </h4>
        </a>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 css-kgdgla-MuiGrid-root"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          data-testid="menu__button"
          tabindex="0"
          type="button"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Menu
          </p>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
